sockeye/data_io.py

```diff
@@ -322,60 +322,72 @@ def shard_data(source_fname: str,
 
     with ExitStack() as exit_stack:
-        source_shards = []
-        target_shards = []
-        # create shard files:
-        for fname in source_shard_fnames:
-            source_shards.append(exit_stack.enter_context(smart_open(fname, mode="wt")))
-        for fname in target_shard_fnames:
-            target_shards.append(exit_stack.enter_context(smart_open(fname, mode="wt")))
-        shards = list(zip(source_shards, target_shards, per_shard_stat_accumulators))
-
-        source_iter = SequenceReader(source_fname, vocab_source, add_bos=False)
-        target_iter = SequenceReader(target_fname, vocab_target, add_bos=True)
-
-        random_shard_iter = iter(lambda: random.choice(shards), None)
-
-        for source, target, (source_shard, target_shard, shard_stats) in zip(source_iter, target_iter,
-                                                                             random_shard_iter):
+        source_shards = [exit_stack.enter_context(smart_open(f, mode="wt")) for f in source_shard_fnames]
+        target_shards = [exit_stack.enter_context(smart_open(f, mode="wt")) for f in target_shard_fnames]
+        source_factor_shards = [
+            [exit_stack.enter_context(smart_open(f, mode="wt")) for f in source_factor_shard_fnames[i]] for i in
+            range(len(source_factor_fnames))]
+
+        source_iter = SequenceReader(source_fname, source_vocab, add_bos=False)
+        target_iter = SequenceReader(target_fname, target_vocab, add_bos=True)
+        source_factor_iters = [SequenceReader(f, v, add_bos=False) for f, v in zip(source_factor_fnames,
+                                                                                   source_factor_vocabs)]
+        random_shard_iter = iter(lambda: random.randrange(num_shards), None)
+
+        for sources, target, random_shard_index in zip(zip(source_iter, *source_factor_iters), target_iter,
+                                                       random_shard_iter):
+            source, *source_factors = sources
+            random_shard_index = cast(int, random_shard_index)
             source_len = len(source)
             target_len = len(target)
 
             buck_idx, buck = get_parallel_bucket(buckets, source_len, target_len)
             data_stats_accumulator.sequence_pair(source, target, buck_idx)
-            shard_stats.sequence_pair(source, target, buck_idx)
+            per_shard_stat_accumulators[random_shard_index].sequence_pair(source, target, buck_idx)
 
             if buck is None:
                 continue
 
-            source_shard.write(ids2strids(source) + "\n")
-            target_shard.write(ids2strids(target) + "\n")
+            source_shards[random_shard_index].write(ids2strids(source) + "\n")
+            target_shards[random_shard_index].write(ids2strids(target) + "\n")
+            for i, line in enumerate(source_factors):
+                source_factor_shards[i][random_shard_index].write(ids2strids(line) + "\n")
```

Comment:
I think it would be good to have a check that the files are indeed token parallel somewhere. Would this be a good place?

Reply:
see my latest commit